statistics/cmsketch.go

```diff
@@ -652,3 +627,57 @@ func (c *TopN) RemoveVal(val []byte) {
 func NewTopN(n int) *TopN {
 	return &TopN{TopN: make([]TopNMeta, 0, n)}
 }
+
+// MergeTopN is used to merge more TopN structures to generate a new TopN struct by the given size.
+// The input parameters are multiple TopN structures to be merged and the size of the new TopN that will be generated.
+// The output parameters are the newly generated TopN structure and the remaining numbers.
+// Notice: If n == 0, we will let n = max(len(TopN.TopN))
+func MergeTopN(topNs []*TopN, n uint32) (*TopN, []TopNMeta) {
+	needTopNNum := false
+	if n == 0 {
+		needTopNNum = true
+	}
+	totCnt := uint64(0)
+	for _, topN := range topNs {
+		totCnt += topN.TotalCount()
+	}
+	if totCnt == 0 {
+		return nil, nil
+	}
+	// Different TopN structures may hold the same value, we have to merge them.
+	counter := make(map[hack.MutableString]uint64)
+	for _, topN := range topNs {
+		if needTopNNum && uint32(len(topN.TopN)) > n {
+			n = uint32(len(topN.TopN))
+		}
+		for _, val := range topN.TopN {
+			counter[hack.String(val.Encoded)] += val.Count
+		}
+	}
+
+	numTop := len(counter)
+	sorted := make([]uint64, numTop)
+	for _, cnt := range counter {
+		sorted = append(sorted, cnt)
+	}
+	sort.Slice(sorted, func(i, j int) bool {
+		return sorted[i] > sorted[j]
+	})
+	n = mathutil.MinUint32(uint32(numTop), n)
+	// lastTopCnt is the smallest value in the new TopN structure
+	lastTopCnt := sorted[numTop-1]
+
+	var finalTopN TopN
+	finalTopN.TopN = make([]TopNMeta, 0, n)
+	popedTopNPair := make([]TopNMeta, 0, uint32(numTop)-n)
+	for value, cnt := range counter {
+		data := hack.Slice(string(value))
+		if cnt >= lastTopCnt {
+			finalTopN.AppendTopN(data, cnt)
```

Comment:
If there are multiple TopN items whose counts are equal to `lastTopCnt`, will the final size of `finalTopN` be larger than the input argument `n`? 
This case is OK for me. I just wonder if this case will happen sometimes.

Reply:
Yes. The final size of `finalTopN` may be larger than the `n`.